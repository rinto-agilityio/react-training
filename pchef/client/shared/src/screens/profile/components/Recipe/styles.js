// Libs
import { StyleSheet } from 'react-native'

// Themes
import { COLORS, METRICS, FONTS } from '../../../../themes'

const styles = StyleSheet.create({
  wrapper: {
    display: 'flex',
  },

  largeWrapper: {
    width: METRICS.smallPercentItem,
    paddingLeft: METRICS.mediumPadding,
    paddingRight: METRICS.mediumPadding,
    minWidth: METRICS.smallScreen,
  },

  // Title
  title: {
    ...METRICS.truncate,
    color: COLORS.black,
    fontWeight: FONTS.fontWeight.large,
    paddingLeft: METRICS.largePadding,
    paddingRight: METRICS.largePadding,
  },
  smallTitle: {
    fontSize: FONTS.fontSize.medium,
  },
  mediumTitle: {
    fontSize: FONTS.fontSize.large,
  },
  largeTitle: {
    fontSize: FONTS.fontSize.extraLarge,
  },

  // Description
  description: {
    color: COLORS.baseGray,
    fontWeight: FONTS.fontWeight.large,
    paddingLeft: METRICS.largePadding,
    paddingRight: METRICS.largePadding,
  },
  smallDescription: {
    fontSize: FONTS.fontSize.small,
  },
  mediumDescription: {
    fontSize: FONTS.fontSize.medium,
  },
  largeDescription: {
    fontSize: FONTS.fontSize.large,
  },

  // Image
  image: {
    width: '100%',
    marginTop: METRICS.mediumMargin,
  },
  smallImage: {
    height: METRICS.smallBgImage,
  },
  mediumImage: {
    height: METRICS.mediumBgImage,
  },
  largeImage: {
    height: METRICS.largeBgImage,
  },
})

export default styles
