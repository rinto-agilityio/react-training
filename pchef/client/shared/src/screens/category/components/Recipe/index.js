// @flow
// add flow above to fix for using flow with React.memo

// Libs
import React, { memo } from 'react'
import { View, Text } from 'react-native'

// Commons
import Wrapper from '../../../../layout/Wrapper'
import Icon from '../../../../components/Icon'
import Image from '../../../../components/Image'

// Constants and Helpers
import { formatStringToArray } from '../../../../helpers/utils'
import { SEPARATOR_SPLIT_STRING, ELLIPSIS_LINE, DEFAULT_IMAGE } from '../../../../constants/index'

// Themes
import { COLORS, FONTS } from '../../../../themes'

// Styles
import styles from './styles'

type Props = {
  recipe: {
    id: string,
    title: string,
    imgUrl: string,
    description: string,
  },
  isGrid: boolean,
  isFavorite: boolean,
  onPressIcon?: (id: string) => void | Promise<void>,
  size?: string,
  color?: string,
  handlePressImage?: () => void,
  customIconStyle: Object,
}

const Recipe = ({
  recipe,
  onPressIcon = () => {},
  size = '',
  color,
  handlePressImage,
  isGrid,
  isFavorite,
  customIconStyle,
}: Props) => {
  const { id, title, imgUrl, description } = recipe
  const formatDescription =
    formatStringToArray(description, SEPARATOR_SPLIT_STRING) || []

  return (
    <View style={[styles.recipe, isGrid ? styles[`${size}GridContent`] : styles[`${size}ListContent`]]}>
      <Text
        numberOfLines={ELLIPSIS_LINE}
        ellipsizeMode="tail"
        style={[styles.title, styles[`${size}Title`]]}
      > {title}
      </Text>
      <Wrapper direction="row" childPosition="left">
        <Image
          url={imgUrl || DEFAULT_IMAGE}
          customImageStyle={[styles.image, styles[`${size}Image`]]}
          handleTouch={handlePressImage}
        />
        <Wrapper
          direction="column"
          childPosition="middle"
          customStyles={[styles.wrapper, isGrid ? styles[`${size}GridWrapper`] : styles[`${size}Wrapper`]]}
        >
          {/* ingredient description */}
          <View style={styles.wrapperText}>
            {formatDescription.map((item, index) => (
              <Text
                numberOfLines={ELLIPSIS_LINE}
                ellipsizeMode="tail"
                key={index}
                style={styles[`${size}Text`]}
              >
                {item}
              </Text>
            ))}
          </View>

          <Icon
            wrapperIconStyle={[styles.icon, customIconStyle]}
            name="favorite"
            size={
              size === 'large'
                ? FONTS.fontSize.extraLarge
                : FONTS.fontSize.medium
            }
            onPress={() => onPressIcon(id)}
            color={isFavorite ? COLORS.red : color}
          />
        </Wrapper>
      </Wrapper>
    </View>
  )
}

Recipe.defaultProps = {
  onPressIcon: () => {},
  size: 'medium',
  color: COLORS.grayDarker,
  handlePressImage: () => {},
}

export default memo<Props>(Recipe)
