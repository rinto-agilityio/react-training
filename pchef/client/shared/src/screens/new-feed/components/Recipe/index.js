// @flow
// add flow above to fix for using flow with React.memo

// Libs
import React, { memo } from 'react'
import { Text, View } from 'react-native'

// Styles
import styles from './styles'

// Components
import Image from '../../../../components/Image'
import Reaction from '../../../../components/Reaction'

// Helpers
import {
  checkContainField,
  formatFiledOnObject,
  checkContain,
} from '../../../../helpers/utils'

// Constants
import { DEFAULT_IMAGE } from '../../../../constants'

import type { RecipeType } from '../../../../types'

type Props = {
  recipe: RecipeType,
  size?: string,
  favoriteRecipe: Array<{
    id: string,
  }>,
  userToggleRecipe: (
    recipeId: string,
    favoriteRecipe: Array<{ id: string }>
  ) => Promise<{ data: { userToggleRecipe: { results: Array<string> } } }>,
  handleClickRecipe: (recipeId: string) => void,
  userToggleVote: (
    recipeId: string,
    votes: Array<string>,
    userId: string
  ) => Promise<{ data: { userToggleVote: { results: Array<string>}}}>,
  userId: string,
  wrapperIconStyle: Object,
}

const Recipe = ({
  recipe,
  size = 'large',
  favoriteRecipe,
  userToggleRecipe,
  handleClickRecipe,
  userToggleVote,
  userId,
  wrapperIconStyle,
}: Props) => {
  const { id, title, description, thumbnail, votes } = recipe

  const isFavorited = checkContainField(favoriteRecipe, id)

  const handleSaveRecipe = async () => {
    await userToggleRecipe(id, favoriteRecipe).then(({ data }) => {
      const {
        userToggleRecipe: { results },
      } = data

      if (results) {
        checkContainField(formatFiledOnObject(results), id)
      }
    })
  }

  const handleToggleVote = async () => {
    await userToggleVote(id, votes, userId)
      .then(({ data }) => {
        const {
          userToggleVote: { results },
        } = data

        if (results) {
          checkContain(votes, id)
        }
      })
  }

  return (
    <View style={[styles.wrapper, styles[`${size}Wrapper`]]}>
      <View style={styles.recipe}>
        <Text
          style={[styles.title, styles[`${size}Title`]]}
          numberOfLines={1}
          ellipsizeMode="tail"
          onPress={() => handleClickRecipe(recipe.id)}
        >
          {title}
        </Text>
        <Text
          style={[styles.description, styles[`${size}Description`]]}
          numberOfLines={1}
          ellipsizeMode="tail"
        >
          {description}
        </Text>
        <Image
          url={thumbnail || DEFAULT_IMAGE}
          customImageStyle={[styles.image, styles[`${size}Image`]]}
          handleTouch={() => handleClickRecipe(recipe.id)}
        />
      </View>
      <Reaction
        votes={votes}
        size={size}
        isFavorited={isFavorited}
        onPressFavorite={handleSaveRecipe}
        onPressVote={handleToggleVote}
        isVote={checkContain(votes, userId)}
        wrapperIconStyle={wrapperIconStyle}
      />
    </View>
  )
}

Recipe.defaultProps = {
  size: 'large',
}

export default memo<Props>(Recipe)
