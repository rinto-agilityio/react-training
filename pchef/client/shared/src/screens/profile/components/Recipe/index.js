// Libs
import React from 'react'
import { Text, View, TouchableOpacity } from 'react-native'

// Styles
import styles from './styles'

// Components
import Image from '../../../../components/Image'
import Reaction from '../../../../components/Reaction'

// Helpers
import { checkContainField, checkContain } from '../../../../helpers/utils'

// Constanst
import { DEFAULT_IMAGE } from '../../../../constants'

type Props = {
  recipe: {
    id: string,
    title: string,
    description: string,
    imgUrl: string,
    votes: Array<string>,
  },
  size?: string,
  favoriteRecipe: Array<{
    id: string,
  }>,
  userToggleRecipe: (
    recipeId: string,
    favoriteRecipe: Array<{
      id: string,
    }>
  ) => Promise<{ data: { userToggleRecipe: { results: Array<string> } } }>,
  userToggleVote: (
    recipeId: string,
    votes: Array<string>,
    userId: string
  ) => Promise<{ data: { userToggleVote: { results: Array<string>}}}>,
  userId: string,
}

const Recipe = ({
  recipe,
  size = 'large',
  favoriteRecipe,
  userToggleRecipe,
  userToggleVote,
  userId,
}: Props) => {
  const { id, title, description, imgUrl, votes } = recipe

  // Call mutation for save recipe
  const handleSaveRecipe = () => {
    userToggleRecipe(id, favoriteRecipe)
  }

  const handleToggleVote = async () => {
    await userToggleVote(id, votes, userId)
      .then(({ data }) => {
        const {
          userToggleVote: { results },
        } = data

        if (results) {
          checkContain(votes, id)
        }
      })
  }

  const isFavorited = checkContainField(favoriteRecipe, id)

  return (
    <TouchableOpacity style={[styles.wrapper, styles[`${size}Wrapper`]]}>
      <View style={styles.recipe}>
        <Text
          style={[styles.title, styles[`${size}Title`]]}
          numberOfLines={1}
          ellipsizeMode="tail"
        >
          {title}
        </Text>
        <Text
          style={[styles.description, styles[`${size}Description`]]}
          numberOfLines={1}
          ellipsizeMode="tail"
        >
          {description}
        </Text>
        <Image
          url={imgUrl || DEFAULT_IMAGE}
          customImageStyle={[styles.image, styles[`${size}Image`]]}
        />
      </View>
      <Reaction
        votes={votes}
        size={size}
        isFavorited={isFavorited}
        isVote={checkContain(votes, userId)}
        onPressFavorite={handleSaveRecipe}
        onPressVote={handleToggleVote}
      />
    </TouchableOpacity>
  )
}

Recipe.defaultProps = {
  size: 'large',
}

export default Recipe
