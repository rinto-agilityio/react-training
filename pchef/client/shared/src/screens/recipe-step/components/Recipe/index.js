// Libs
import React, { useEffect, useState } from 'react'
import { Text, View } from 'react-native'

// Styles
import styles from './styles'

// Themes
import { COLORS, METRICS } from '../../../../themes'

// Components
import Progress from './ProgressStep'
import ImageBackground from '../../../../components/ImageBackground'
import Reaction from '../../../../components/Reaction'
import Comment from '../../../recipe/components/Comment'
import Loading from '../../../../components/Loading'
import Error from '../../../../components/Error'

// mock data
import { recipes } from '../../../../mocks'

// utils
import { findStep, compareStep, customError } from '../../../../helpers/utils'

type Props = {
  recipeSteps: Array<{
    description: string,
    imgUrl: string,
    step: number,
    title: string,
  }>,
  recipe: {
    title: string,
    votes: Array<number>,
    subTitle: string,
    views: number,
    userId: number,
    steps: Array<{
      step: number
    }>,
  },
  size: string,
  onPress?: () => void,
  customRecipe?: {},
  customTitle?: {},
  customDescription?: {},
  customImage?: {},
  customSubTitle?: {},
  customTitleStep?: {},
  loading: boolean,
  error: string
}

const Recipe = ({
  size = 'large',
  onPress,
  customRecipe,
  customTitle,
  customDescription,
  customImage = {},
  customSubTitle,
  customTitleStep,
  recipeSteps = [{
    description: '',
    imgUrl: '',
    step: 1,
    title: '',
  }],
  loading,
  error,
}: Props) => {
  // order recipeSteps by step asc
  const orderRecipeSteps = recipeSteps.sort(compareStep)
  const defaultStepInfo = orderRecipeSteps[0]
  const [stepInfo, setStepInfo] = useState()

  useEffect(() => (
    setStepInfo(defaultStepInfo)
  ), [loading, defaultStepInfo])

  const {
    title,
    subTitle,
    votes,
    userId,
    views,
  } = recipes[0]

  if (loading) {
    return <Loading />
  }

  if (error) {
    return <Error message={customError(error.graphQLErrors)} />
  }
  /**
   * Handle when user click prev or next icon
   * param {name}
   */
  const onPressSelectStep = name => {
    let nextStepInfo

    switch (name) {
      case 'next':
        nextStepInfo = findStep(orderRecipeSteps, stepInfo.step + 1)
        break;
      case 'prev':
        nextStepInfo = findStep(orderRecipeSteps, stepInfo.step - 1)
        break;
      default:
        break;
    }
    setStepInfo(nextStepInfo)
  }

  /**
   * Handle When user click any step
   * @param {step}
   */
  const onPressStep = step => {
    const stepInfoSelect = findStep(orderRecipeSteps, step)
    setStepInfo(stepInfoSelect)
  }

  return (
    <View style={[styles.wrapper, styles[`${size}Wrapper`]]}>
      <View
        style={[
          styles.recipe,
          customRecipe,
        ]}
      >
        <Text
          style={[
            styles.title,
            styles[`${size}Title`],
            customTitle,
          ]}
          onPress={onPress}
        >
          {title}
        </Text>
        <Text
          style={[
            styles.title,
            styles[`${size}Description`],
            customSubTitle,
          ]}
          onPress={onPress}
        >
          {subTitle}
        </Text>
        <ImageBackground
          url={stepInfo.imgUrl}
          customImageBg={[
            styles.image,
            styles[`${size}Image`],
            customImage,
          ]}
          resizeMode="cover"
        >
          <Text>
            <Text
              style={[
                styles.titleStep,
                styles[`${size}Description`],
                customTitleStep,
              ]}
            >
              {stepInfo.title}
            </Text>
          </Text>
          <Progress
            steps={orderRecipeSteps}
            size={size}
            step={stepInfo.step}
            onPressSelectStep={onPressSelectStep}
            onPressStep={onPressStep}
          />
        </ImageBackground>
      </View>
      <Text
        style={[
          styles.description,
          styles[`${size}Description`],
          customDescription,
        ]}
      >
        {stepInfo.description}
      </Text>
      <Reaction
        votes={votes}
        size={size}
        isFavorited={false}
      />
      <Comment
        name={`by ${userId}`}
        publishedAt={Date.now()}
        isGetTime
        customStyle={{
          paddingLeft: METRICS.largePadding,
        }}
        customNameStyles={{
          fontWeight: 'bold',
        }}
        customContentStyles={{
          fontSize: 12,
          color: COLORS.baseGray,
        }}
      />
      <Text
        style={[
          styles.views,
          styles[`${size}Views`],
        ]}
      >
        {`viewed by ${views}`}
      </Text>
    </View>
  )
}

Recipe.defaultProps = {
  onPress: () => {},
  customRecipe: {},
  customTitle: {},
  customDescription: {},
  customSubTitle: {},
  customTitleStep: {},
  customImage: {},
}

export default Recipe
