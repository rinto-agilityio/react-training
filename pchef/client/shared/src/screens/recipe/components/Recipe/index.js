// Libs
import React, { useState } from 'react'
import { View } from 'react-native'

// Styles
import styles from './styles'

// Components
import Ingredients from './Ingredients'
import Directions from './Directions'
import Loading from '../../../../components/Loading'
import Modal from '../../../../components/Modal'
import Error from '../../components/Error'

// Helper
import { customError } from '../../../../helpers/utils'

type Props = {
  getRecipe: {
    description: string,
    id: string,
  },
  size: string,
  onSelectStep: (id: string) => void,
  loading: boolean,
  error: Object,
  recipeSteps: Array<{
    step: number,
    title: string
  }>,
  history: Object,
}

const Recipe = ({
  getRecipe = {
    description: '',
    id: '',
  },
  size = 'medium',
  onSelectStep,
  loading,
  error,
  recipeSteps = [{
    step: 1,
    title: '',
  }],
  history,
}: Props) => {
  const [visible, setVisible] = useState(true)

  if (loading) {
    return <Loading />
  }

  const handleNavigateLogin = () => {
    setVisible(false)
    history.push('/login')
  }

  if (error) {
    return (
      <Modal
        visible={visible}
        onDismiss={() => handleNavigateLogin()}
        onSubmit={() => handleNavigateLogin()}
        size={size}
      >
        <Error message={customError(error.graphQLErrors)} size="medium" />
      </Modal>
    )
  }

  const {
    description,
    id,
  } = getRecipe

  return (
    <View style={[styles.wrapper, styles[`${size}Wrapper`]]}>
      <Ingredients
        description={description}
        size={size}
      />
      <Directions
        steps={recipeSteps}
        size={size}
        onSelectStep={() => onSelectStep(id)}
      />
    </View>
  )
}

export default Recipe
