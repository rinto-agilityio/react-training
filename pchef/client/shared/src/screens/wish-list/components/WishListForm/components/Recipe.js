// Libs
import React from 'react'
import { Text, View } from 'react-native'

// Styles
import styles from '../styles'

// Components
import Image from '../../../../../components/Image'

// Constants
import { DEFAULT_IMAGE } from '../../../../../constants'

type Props = {
  recipe: {
    title: string,
    imgUrl: string,
  },
  size?: string,  
}

const Recipe = ({
  recipe,
  size = 'large',
}: Props) => {
  const { title, imgUrl } = recipe

  return (
    <View style={[styles.wrapper, styles[`${size}Wrapper`]]}>
      <View style={styles.recipe}>
        <Text
          style={[styles.title, styles[`${size}Title`]]}
          numberOfLines={1}
          ellipsizeMode="tail"
        >
          { title }
        </Text>
        <Image
          url={imgUrl || DEFAULT_IMAGE}
          customImageStyle={[styles.image, styles[`${size}Image`]]}
        />
      </View>
    </View>
  )
}

export default Recipe
