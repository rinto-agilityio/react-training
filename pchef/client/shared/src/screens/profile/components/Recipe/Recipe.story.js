// Libs
import React from 'react'
import { storiesOf } from '@storybook/react'

// Helpers
import { recipes } from '../../../../mocks'

// Components
import Recipe from '.'

storiesOf('Profile recipe', module)
  .add('medium', () => (
    <Recipe size="medium" recipe={recipes[0]} favoriteRecipe={['idtest']} />
  ))
  .add('large', () => (
    <Recipe size="large" recipe={recipes[0]} favoriteRecipe={['idtest']} />
  ))
