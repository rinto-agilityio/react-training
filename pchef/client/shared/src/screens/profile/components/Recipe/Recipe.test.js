// Helpers
import { recipes } from '../../../../mocks'
import { DEFAULT_IMAGE } from '../../../../constants'

// Components
import Recipe from '.'
import Reaction from '../../../../components/Reaction'
import Image from '../../../../components/Image'

describe('Recipe new feed', () => {
  const props = {
    recipe: recipes[0],
    favoriteRecipe: ['idtest'],
    userToggleRecipe: jest.fn(),
    userToggleVote: jest.fn(),
  }

  let component

  beforeEach(() => {
    component = shallow(<Recipe {...props} />)
  })

  it('Renders correctly recipe commponent', () => {
    expect(component).toMatchSnapshot()
  })

  it('Render recipe component with size medium', () => {
    component.setProps({
      size: 'medium',
    })
    expect(component).toMatchSnapshot()
  })

  it('Should render recipe with default image', () => {
    component.setProps({
      recipe: {
        ...recipes[0],
        imgUrl: null,
      },
    })

    expect(component.find(Image).props().url).toEqual(DEFAULT_IMAGE)
  })

  it('Should call userToggleVote when press on vote', () => {
    const reaction = component.find(Reaction)
    reaction.props().onPressVote()

    expect(props.userToggleVote).toHaveBeenCalled()
  })

  it('Should call userToggleRecipe when press on favorite', () => {
    const reaction = component.find(Reaction)
    reaction.props().onPressFavorite()

    expect(props.userToggleRecipe).toHaveBeenCalled()
  })
})
