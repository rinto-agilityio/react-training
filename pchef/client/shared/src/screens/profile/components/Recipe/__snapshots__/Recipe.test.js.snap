// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Recipe new feed Render recipe component with size medium 1`] = `ShallowWrapper {}`;

exports[`Recipe new feed Renders correctly recipe commponent 1`] = `ShallowWrapper {}`;
